File: examples/ospf-kotlin-example/src/main/fuookami/ospf/kotlin/example/framework_demo/demo4/rule_context/model/Restriction.kt
package fuookami.ospf.kotlin.example.framework_demo.demo4.rule_context.model

import fuookami.ospf.kotlin.utils.math.*
import fuookami.ospf.kotlin.example.framework_demo.demo4.infrastructure.*
import fuookami.ospf.kotlin.example.framework_demo.demo4.flight_task_context.model.*

enum class RestrictionType {
    Weak,
    ViolableStrong,
    Strong,
}

sealed interface Restriction {
    val type: RestrictionType

    fun related(aircraft: Aircraft): Boolean

    fun check(task: FlightTask): RestrictionCheckingResult
    fun check(task: FlightTask, aircraft: Aircraft): RestrictionCheckingResult
    fun check(task: FlightTask, recoveryPolicy: FlightTaskAssignment): RestrictionCheckingResult
}

sealed interface RestrictionCheckingResult {
    val restriction: Restriction

    val type get() = restriction.type
}

data class NotMatter(
    override val restriction: Restriction
) : RestrictionCheckingResult

data class Violate(
    override val restriction: Restriction
) : RestrictionCheckingResult

data class NotViolate(
    override val restriction: Restriction
) : RestrictionCheckingResult

data class ViolableViolate(
    override val restriction: Restriction
) : RestrictionCheckingResult

enum class RelationRestrictionCategory {
    BlackList,
    WhiteList
}

class RelationRestriction(
    override val type: RestrictionType,
    val category: RelationRestrictionCategory,
    val dep: Airport,
    val arr: Airport,
    val aircrafts: Set<Aircraft>,
    val weight: Flt64 = Flt64.one,
    val cost: Flt64? = null
) : Restriction {
    override fun related(aircraft: Aircraft): Boolean {
        return aircrafts.contains(aircraft)
    }

    override fun check(task: FlightTask): RestrictionCheckingResult {
        assert(task.isFlight)
        return check(task.dep, task.arr, task.aircraft)
    }

    override fun check(task: FlightTask, aircraft: Aircraft): RestrictionCheckingResult {
        assert(task.isFlight)
        return check(task.dep, task.arr, aircraft)
    }

    override fun check(task: FlightTask, recoveryPolicy: FlightTaskAssignment): RestrictionCheckingResult {
        assert(task.isFlight)
        val dep = recoveryPolicy.route?.dep ?: task.dep
        val arr = recoveryPolicy.route?.arr ?: task.arr
        val aircraft = recoveryPolicy.aircraft ?: task.aircraft
        return check(dep, arr, aircraft)
    }
    
    private fun check(dep: Airport, arr: Airport, aircraft: Aircraft?): RestrictionCheckingResult {
        if (dep != this.dep || arr != this.arr) {
            return NotMatter(this)
        }
        return aircraft?.let { dump(aircrafts.contains(aircraft)) } ?: NotMatter(this)
    }

    private fun violated(hit: Boolean): Boolean {
        return when (category) {
            RelationRestrictionCategory.BlackList -> hit
            RelationRestrictionCategory.WhiteList -> !hit
        }
    }

    private fun dump(hit: Boolean): RestrictionCheckingResult {
        return if (violated(hit)) {
            ViolableViolate(this)
        } else {
            NotViolate(this)
        }
    }
}

File: ospf-rust/ospf-rust-example/src/framework/demo4/domain/rule/model/restriction.rs
//! 限制规则模型 / Restriction rule model

/// 限制类型 / Restriction type
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum RestrictionType {
    /// 通用限制 / General restriction
    General,
    /// 关联限制 / Relation restriction
    Relation,
}

/// 限制检查结果 / Restriction checking result (对齐 Kotlin RestrictionCheckingResult)
#[derive(Debug, Clone)]
pub enum RestrictionCheckingResult {
    /// 无关 / Not applicable
    NotMatter,
    /// 违规，包含原因 / Violation with reason
    Violate { reason: String },
    /// 未违规 / No violation
    NotViolate,
    /// 可豁免违规，包含原因 / Waivable violation with reason
    ViolableViolate { reason: String },
}

/// 限制 / Restriction (对齐 Kotlin Restriction sealed interface)
#[derive(Debug, Clone)]
pub struct Restriction {
    /// 限制标识 / Restriction identifier
    pub id: String,
    /// 限制类型 / Restriction type
    pub restriction_type: RestrictionType,
    /// 限制描述 / Restriction description
    pub description: String,
}

impl Restriction {
    /// 检查限制 / Check restriction
    /// 对齐 Kotlin Restriction.check
    pub fn check(&self) -> RestrictionCheckingResult {
        RestrictionCheckingResult::NotMatter
    }
}
